coverity scan fixes

fabbione opened the pull request:
this PR addresses all the major issues reported by coverity scan on the runtime code.

There are still a bunch of issues left in the test suite and doxyxml.c that is executed only during build time to generate man pages.

coverity scan will report an error in libknet/thread_tx.c related to the channel variable. The error is correct, there is some code that will need to be fixed, but at the moment, the if () branch that sets channel = -1 is dead and nothing in knet triggers that code (it was left in preparation for knet 2.0).

This PR fixes https://github.com/kronosnet/kronosnet/issues/237

fabbione committed:
[sctp] fix deference after null check

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[sctp] free access list only if the socket is valid

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[rx] align data types

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[handle] make sure that the pmtud buf contains at least knet header size

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[sctp] revalidate fd to make coverity scan happy

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[tests] fix knet_bench coverity errors

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[common] fix dlopen error handling

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[rx] better error report if we can't resolve hostname / port

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[acl] avoid forward null deferencing

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[sctp] cleanup bugs detected in error paths by coverity scan

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[nozzle] fix negative return detected by coverity scan

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[nozzle] don't leak memory on error

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[compress] don't leak memory in case of errors during zstd init

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[logging] make sure not to overrun buffers by pre-allocating them

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[compress] do not overrun allocated array for compress modules

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione commented:
retest this please

fabbione commented:
retest this please

fabbione commented:
retest this please

jfriesse commented on libknet/compress.c:
```diff
@@ -32,7 +32,7 @@
  */
 
-static compress_model_t compress_modules_cmds[] = {
+static compress_model_t compress_modules_cmds[KNET_MAX_COMPRESS_METHODS + 1] = {
```
Are you sure this is really needed? The array will autoinitialize because of last line

fabbione replied:
according to coverity scan the size of the array is the number of entries in the array, so 12 atm. When looping around it complains that it might be possible access out-of-boundaries.
Pre-allocating the array will make looping safer.

jfriesse replied:
But the array is preallocated because of last line in the array initializer.

jfriesse replied:
This was discussed on IRC. Part of the problem is coverity cleverness (removing idx < KNET_MAX_COMPRESS_METHODS would work same way).

jfriesse commented on libknet/internals.h:
```diff
@@ -90,8 +90,8 @@ struct knet_host_defrag_buf {
-	uint16_t frag_size;		/* normal frag size (not the last one) */
-	uint16_t last_frag_size;	/* the last fragment might not be aligned with MTU size */
+	ssize_t frag_size;		/* normal frag size (not the last one) */
+	ssize_t last_frag_size;		/* the last fragment might not be aligned with MTU size */
```
whitespace addition?

fabbione replied:
I see two tabs after the ; on both lines.....

jfriesse replied:
Yep, GH.

jfriesse commented on libknet/links_acl_ip.c:
```diff
@@ -279,10 +279,15 @@ int ipcheck_addip(void *fd_tracker_match_entry_head, int index,
 			 */
-			while ((match_entry->next) || (i < index)) {
+
+			while (match_entry->next) {
```
Wouldn't `while ((match_entry->next) && (i < index)) {` (or maybe `(i <= index)`, I'm not sure what is the intended result) work same way?

fabbione replied:
I have tried different combinations including the ones you suggested and somehow it was failing to pass the test suite because it was adding the new ACL in the wrong place.

This code is not on the critical performance path either, so i just decided to keep it simpler and loop only against one variable.

jfriesse replied:
Ok, so it's because index is -1. I don't agree that code is really simpler (it would be simpler with list abstraction, as for example libqb list, queue.h, ...), but this seems to work too.

jfriesse commented on libknet/logging.c:
```diff
@@ -20,7 +20,7 @@
 #include "threads_common.h"
 
-static struct pretty_names subsystem_names[] =
+static struct pretty_names subsystem_names[KNET_MAX_SUBSYSTEMS] =
```
Are you sure this shouldn't be + 1 (as with KNET_LOG_DEBUG)?

fabbione replied:
MAX_SUBSYSTEMS has already +1 included.

jfriesse commented on libknet/transport_sctp.c:
```diff
@@ -303,7 +303,9 @@ int sctp_transport_tx_sock_error(knet_handle_t knet_h, int sockfd, int recv_err,
 			usleep(knet_h->threads_timer_res / 16);
-			pthread_rwlock_rdlock(&knet_h->global_rwlock);
+			if (pthread_rwlock_rdlock(&knet_h->global_rwlock) != 0) {
+				log_warn(knet_h, KNET_SUB_TRANSP_SCTP, "Unable to get read lock, results might be unpredictable!");
```
Is logging warning really best (or only) thing to do? not-getting lock is just asking for a trouble, so maybe it may make sense to return error code?

fabbione replied:
yeah you are right, it might be possible to return an error. I'll look into it. thanks!

fabbione commented:
On 19/07/2019 10.07, Jan Friesse wrote:
> *@jfriesse* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In libknet/compress.c
> <https://github.com/kronosnet/kronosnet/pull/240#discussion_r305244622>:
> 
>> @@ -32,7 +32,7 @@
>   * Always add new items before the last NULL.
>   */
>  
> -static compress_model_t compress_modules_cmds[] = {
> +static compress_model_t compress_modules_cmds[KNET_MAX_COMPRESS_METHODS + 1] = {
> 
> But the array is preallocated because of last line in the array initializer.

apparently not, we had a similar patch in the past from Suse to fix
another array. I think gcc has gotten smarter and only allocate by
entries in the array.

Fabio

> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kronosnet/kronosnet/pull/240?email_source=notifications&email_token=AAC3VLKADJMOE7GIDQHF7VDQAFY4RA5CNFSM4IE3D4Q2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB66VVZQ#discussion_r305244622>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC3VLOSQEEKOZCT5PVLINLQAFY4RANCNFSM4IE3D4QQ>.
>

fabbione commented:
On 19/07/2019 10.06, Jan Friesse wrote:
> *@jfriesse* commented on this pull request.
> 
> ------------------------------------------------------------------------
> 
> In libknet/links_acl_ip.c
> <https://github.com/kronosnet/kronosnet/pull/240#discussion_r305244100>:
> 
>> @@ -279,10 +279,15 @@ int ipcheck_addip(void *fd_tracker_match_entry_head, int index,
>  			/*
>  			 * find the end of the list or stop at "index"
>  			 */
> -			while ((match_entry->next) || (i < index)) {
> +
> +			while (match_entry->next) {
> 
> Ok, so it's because index is -1. I don't agree that code is really
> simpler (it would be simpler with list abstraction, as for example libqb
> list, queue.h, ...), but this seems to work too.

We could switch the whole code to use that but to be honest i am not
even sure it's worth the effort. I think we have one or two lists in the
whole code base as we mostly use arrays to point to data.

Fabio

> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/kronosnet/kronosnet/pull/240?email_source=notifications&email_token=AAC3VLPCMGBHOPVQHK4XODDQAFYW3A5CNFSM4IE3D4Q2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB66VQLY#discussion_r305244100>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAC3VLJZHLO6NZ6RQALRCXDQAFYW3ANCNFSM4IE3D4QQ>.
>

fabbione committed:
[coverity] add test targets to run coverity automatically

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione commented:
@jfriesse i think the patch is good, I would probably double check the last commit since it's a bit more invasive that i like

fabbione committed:
[PMTUd] do not double unlock global read lock

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[nozzle] avoid tons of possible buffer overruns

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[test] simplify flush log

allocate on stack only once and make sure strings are null terminated
drop useless read loop since log msg are always smaller than PAGE_SIZE
and read are atomic at that level

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[common] make sure string is null terminated

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[tx] drop unnecessary usleep when sending to localhost

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[nozzle] fix a few coverity errors in the test suite

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

fabbione committed:
[tx] clean up channel management code for internal communications

the code is still not in use but it's more clear and doesn't trigger
memory overrun

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

chrissie-c commented on libknet/transport_sctp.c:
```diff
@@ -274,6 +274,25 @@ static int _create_connect_socket(knet_handle_t knet_h, struct knet_link *kn_lin
 	return err;
 }
 
+static void _unlock_relock(knet_handle_t knet_h)
```
It's an oddly undescriptive name. How about _lock_sleep_relock() ?

fabbione commented:
retest this please

chrissie-c reviewed (dismissed):
ACK. looks sensible to me and tests OK.

fabbione committed:
[sctp] retry locking in case of failure

Signed-off-by: Fabio M. Di Nitto <fdinitto@redhat.com>

jfriesse reviewed (approved):
Looks good now so ACK.

fabbione commented:
retest this please

fabbione commented:
retest this please
